java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,643 @@
+      this.tree = tree;
+      this.semantic = semantic;
+    }
+
+    public CompilationUnitTree tree() {
+      return tree;
+    }
+
+    public SemanticModel semantic() {
+      return semantic;
+    }
+  }
+
+  private final ActionParser<Tree> parser = JavaParser.createParser();
+
+  public ScannedFile scan(File source, ClassLoader classLoader) {
+    CompilationUnitTree tree = (CompilationUnitTree) parser.parse(source);
+    SemanticModel model = SemanticModel.createFor(tree, classLoader);
+
+    return new ScannedFile(tree, model);
+  }
+
+  public static class TaintSummary {
+
+    private final Set<TaintSource> result;
+    private final Set<DirectTaintSource> methodCalls;
+
+    public TaintSummary(Set<TaintSource> result, Set<DirectTaintSource> methodCalls) {
+      this.result = result;
+      this.methodCalls = methodCalls;
+    }
+
+    public boolean resultCanBeTaintedBy(String s) {
+      for (TaintSource ts: result) {
+        if (ts.toString().equals(s)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    public int resultTaintedPaths() {
+      return result.size();
+    }
+
+    public boolean callsMethod(String s) {
+      for (DirectTaintSource ts: methodCalls) {
+        if (ts.toString().equals(s)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    public int methodCalls() {
+      return methodCalls.size();
+    }
+
+  }
+
+  public interface TaintSource {
+
+    boolean canBeTainted();
+    TaintSource unionWith(TaintSource other);
+
+  }
+
+  /**
+   * Literals
+   * Non-string operands
+   */
+  public static class TaintFreeSource implements TaintSource {
+
+    @Override
+    public boolean canBeTainted() {
+      return false;
+    }
+
+    @Override
+    public TaintSource unionWith(TaintSource other) {
+      return other;
+    }
+
+    @Override
+    public String toString() {
+      return "taint-free";
+    }
+
+    @Override
+    public int hashCode() {
+      return 0;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
+        return false;
+      }
+
+      return true;
+    }
+
+  }
+
+  /**
+   * Read-from method parameter
+   * Read-from field
+   * Read-from method-call return value
+   */
+  public static class DirectTaintSource implements TaintSource {
+
+    private final String signature;
+    private final List<TaintSource> arguments;
+
+    public DirectTaintSource(String signature, List<TaintSource> arguments) {
+      this.signature = signature;
+      this.arguments = arguments;
+    }
+
+    @Override
+    public boolean canBeTainted() {
+      return true;
+    }
+
+    @Override
+    public TaintSource unionWith(TaintSource other) {
+      return IndirectTaintSource.union(this, other);
+    }
+
+    @Override
+    public String toString() {
+      return signature + (arguments.isEmpty() ? "" : "(" + Joiner.on(", ").join(arguments) + ")");
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Add parentheses to make the operator precedence explicit. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864)